docs/ardesia.1.in
.TH ARDESIA "1" "February 2011" "Ardesia @VERSION@; the free digital sketchpad" "Ardesia Manual Pages"

.SH NAME
Ardesia \-  the free digital sketchpad

.SH SYNOPSIS
.B ardesia
[\fIoptions\fR] [\fIfilename\fR]

.SH DESCRIPTION
.PP
Ardesia is the \fIfree digital sketchpad\fP. It  is useful to make coloured free-hand annotations with digital ink everywhere, record it and share on the network. You can draw upon the desktop or import an image and annotate it and redistribute your work to the world.
.PP
Thanks to Ardesia you are free to open any application and fix your ideas and notes. You can use the tool to make effective on-screen presentation, highlight things or point out things of interest. The tool facilitates the online presentations and demos showing in real time your computer screen to anyone in the network. 
.PP
The tool facilitates the online presentations and demos showing in real time your computer screen to anyone in the network. You can use this tool to enhance your lessons or courses working with your preferred applications and your preferred operating system. Create nice tutorial and demos saving the desktop images with your free hand annotations.
.PP
Ardesia uses the interactive Whiteboard Common File (iwb) an native document format, it can export images into several standard image formats.
.SH OPTIONS
Ardesia accepts the following options:
.TP 8
.B  \-h, \-\-help@
Show Ardesia command\-line options.
.TP 8
.B  \-V, \-\-verbose
Enable verbose mode to see the logs
.TP 8
.B  \-d, \-\-decorate
Decorate the window with the borders
.TP 8
.B  \-g, \-\-gravity
Set the gravity of the bar.
Possible values are: east [default], west, north, south

.SH SUGGESTIONS AND BUG REPORTS
Any bugs found should be reported to the online bug-tracking system
available on the web at http://code.google.com/p/ardesia/issues/list. Before reporting bugs, please check to see if the bug has already been reported.

When reporting Ardesia bugs, it is important to include a reliable way to
reproduce the bug, version number of Ardesia (and probably GTK+), OS name
and version, and any relevant hardware specs. If a bug is causing a
crash, it is very useful if a stack trace can be provided. And of
course, patches to rectify the bug are even better.

.SH OTHER INFO
The canonical place to find Ardesia downloads and info is at http://www.cooglecode.com/ardesia. Here you can find links, data sets, mailing list archives and more.

.SH AUTHORS
Pietro Pilolli

.SH COPYRIGHT AND LICENSE
Copyright (C) 2009–2010 by Authors.
Ardesia is free software; you can redistribute it and/or modify it under the terms of the GPL.



NEWS
Ardesia free digital sketchpad


---------------
 ardesia 0.7
---------------

Internationalization:
- Added german translation

Packaging
- New free windows installer created with an inno setup script

Usability
- The default folder where store the created files
  called ardesia workspace now is located under the document folder
- Created a shortcut on desktop that links to the ardesia workspace
- Now to save an image you must push on the the camera photo icon

Features
- Now you can open, save and reuse projects thanks to the
  Interactive Whiteboard Common File Format (iwbcff);
  the artefacts will be saved under the ardesia workspace at the end 
  of the session
- At the end of the program you will be invited to send by email the
  artefacts created in session (iwb file, png screenshots and pdf)
- The virtual keyboard is shown each time that is focused an editable item;
  this feature works only in Linux with the gnome environment and
  in Windows 7 if you have the "Tablet PC Input Panel"
- New crash report dialog
- Improvements in the shape recognizer

Bug Fix:
- Fix in the filler a strange behaviour when you try to fill a not
  closed path shape


---------------
 ardesia 0.6
---------------

Internationalization:
- Add french translation

Usability:
- New text insertion tool to improve the usability;
  implemented a new text insertion cursor, 
  when you click on the page a new cursor blinks on
  the text focus
- New drawing mode tool: you can select the free hand mode,
  the roundifier or the rectifier
- insertion tool integration with the "Tablet PC Input Panel" on windows
- insertion tool integration with the "Florence Ramble Virtual Keyboard"
  on Linux in the Gnome Desktop environment
- vlc used to perform the recording is auto detect if it is
  installed in a default location


Features:
- Ardesia stay above all windows; you can open a program
  in fullscreen and annotate on it
- When a fault occurs it logs the backtrace in tour TMP folder
- Email support
- Pdf email sender; when you quit the program if you have made
  a pdf Ardesia opens your default mailer with a new stub mail
  and it attaches your artefact
- Bug Tracking: If the program is able to found a misbehaviour
  it open an email stub with the log to be sent to the ardesia 
  developers staff
- Florence virtual keyboard integration: if you have it
  installed, will be show/hide when you select the text
  tool (linux only)


---------------
 ardesia 0.5
---------------

Dependencies:
- gtk 2.16.6 included inside the windows package

Usability:
- performance improvement
- faster pdf export
- enhancement in the pressure support

Features:
- Ardesia now works with Kwin and xcompmgr
  composite managers also
- Better shape recognition with the roundifier tool;
  now a new bezier algorithm and ellipse detection
  has been implemented
- low memory usage; the save point are stored on files
- add system that print the backtrace in the case that
  a segmentation fault occurs (linux only)

Bug Fix:
- fix the sound recording on Linux using pulse audio
- fix a bug that caused the conjunction of the new drawn point
  with the last point belong to the old line;
  this problem occurred only using the xtest virtaul device
  (this is used by the python-whiteboard wiimote driver also)
- fix the opening image file as ardesia background right clicking
  on file icon end selecting "open with ardesia"
- fix the missing savepoint after the text insertion that
  caused a wrong behaviour with undo and redo when was used
  the text insertion tool


---------------
 ardesia 0.4
---------------

Dependencies:
- removed the gconf dependency
- removed the ezstream dependency
- require vlc to record and forward the
  live screencast to the icecast server

Usability:
- added Spanish translation

Features:
- Implemented cross platform recorder and stream forwarder
- Added the pen pressure support  for the drawing tablet
- Added the eraser support for the drawing tablet on Linux
- Export as pdf

Bug Fix:
- Fix the file name suggestion in the
  record and save wizard on windows
- Solve the bug in the hide/show button
  on windows
- Fix the colour of the pen cursor synchronization
  on windows


---------------
 ardesia 0.3
---------------

Dependencies:
- ezstream to forward the live screencast
  to the icecast server

Features:
- update to gpl-3
- using getopt; the Unix standard for the command line parsing
- use intltoolize to support the language translation
- added info button with author info and license agreement
  (issue 11)
- added the verbosity option to trace the tool
- undo/redo using vector graphics surfaces instead of raw images
- added electronic whiteboard support
- added tablet support
- add stream forward support to an icecast server
  to make a live online presentation
- now you can different ardesia instances
  one for each workspace
- set the ardesia windows to stay always on the top
- build on windows platform

Bug fix:
- fill now works after a close path drawn
  with or within the highlighter selected
- fix the behaviour of recorder
  Now if the recordmydesktop tool
  is missed a dialog shows the problem
  (issue 9)
- solved some memory leaks
- solve a critical bug that froze the cursor
  when the window loose the focus
- solve a bug that froze the application
  when the roundifier was enabled
  and short a lines was drawn

Usability:
- better user interface
- better arrow painting
- added unlock button
- increase the painting performance
- enhancement in the shape recognizer
- Dynamic change of eraser and pen cursor
  depending on the selected width
- hide the cursor after a paint to avoid that the cursor
  overlap the annotations
- lighter recording using subsampling; this allow to have
  a good video for busier machines too
- Added text insertion using keyboard
- Added text insertion using external text input program
  This feature is tested using
  - the "onboard" virtual keyboard
  - the "florence-ramble" virtual keyboard

src/ardesia.c
/* 
 * Ardesia -- a program for painting on the screen
 * with this program you can play, draw, learn and teach
 * This program has been written such as a freedom sonet
 * We believe in the freedom and in the freedom of education
 *
 * Copyright (C) 2009 Pilolli Pietro <pilolli@fbk.eu>
 *
 * Ardesia is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Ardesia is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


#include <utils.h>
#include <annotation_window.h>
#include <background_window.h>
#include <project_dialog.h>
#include <share_confirmation_dialog.h>
#include <bar.h>


#ifdef HAVE_BACKTRACE
#  include <glibc_backtrace.h>
#endif

#ifdef _WIN32
#  include <windows_backtrace.h>
#endif


/* Print the version of the tool and exit. */
static void 
print_version ()
{
  g_printf ("Ardesia %s; the free digital sketchpad\n\n", PACKAGE_VERSION);
  exit (EXIT_SUCCESS);
}


/* Print the command line help. */
static void
print_help ()
{
  gchar *year = "2009-2010";
  gchar *author = "Pietro Pilolli";
  g_printf ("Usage: %s [options] [filename]\n\n", PACKAGE_NAME);
  g_printf ("Ardesia the free digital sketchpad\n\n");
  g_printf ("options:\n");
  g_printf ("  --verbose ,\t-V\t\tEnable verbose mode to see the logs\n");
  g_printf ("  --decorate,\t-d\t\tDecorate the window with the borders\n");
  g_printf ("  --gravity ,\t-g\t\tSet the gravity of the bar. Possible values are:\n");
  g_printf ("  \t\t\t\teast [default]\n");
  g_printf ("  \t\t\t\twest\n");
  g_printf ("  \t\t\t\tnorth\n");
  g_printf ("  \t\t\t\tsouth\n");
  g_printf ("  --help    ,\t-h\t\tShows the help screen\n");
  g_printf ("  --version ,\t-v\t\tShows version information and exit\n");
  g_printf ("\n");
  g_printf ("filename:\t  \t\tThe interactive Whiteboard Common File (iwb)\n");
  g_printf ("\n");
  g_printf ("%s (C) %s %s\n", PACKAGE_STRING, year, author);
  exit (EXIT_SUCCESS);
}


#ifndef _WIN32
/* Call the dialog that inform the user to enable a composite manager. */
static void
run_missing_composite_manager_dialog ()
{
  GtkWidget *msg_dialog;
  msg_dialog = gtk_message_dialog_new (NULL,
				       GTK_DIALOG_MODAL,
				       GTK_MESSAGE_ERROR,
				       GTK_BUTTONS_OK,
				       gettext ("In order to run Ardesia you need to enable a composite manager"));

  gtk_dialog_run (GTK_DIALOG (msg_dialog));

  if (msg_dialog != NULL)
    {
      gtk_widget_destroy (msg_dialog);
      msg_dialog = NULL;
    }

  exit (EXIT_FAILURE);
}


/* Check if a composite manager is active. */
static void
check_composite_manager (GdkScreen *screen)
{
  gboolean composite = gdk_screen_is_composited (screen);

  if (!composite)
    {
      /* start the enable composite manager dialog. */
      run_missing_composite_manager_dialog ();
    }

}
#endif


/* Set the best colormap available on the system. */
static void
set_the_best_colormap ()
{
  GdkDisplay *display = gdk_display_get_default ();
  GdkScreen  *screen  = gdk_display_get_default_screen (display);
  GdkVisual  *visual  = NULL;
  GdkColormap *colormap = gdk_screen_get_rgba_colormap (screen);

  /* In FreeBSD operating system you might have a composite manager. */
#if ( defined (__freebsd__) || defined (__freebsd) || defined (_freebsd) || defined (freebsd) )
  check_composite_manager (screen);
#endif

  /* In linux operating system you must have a composite manager. */
#ifdef linux
  check_composite_manager (screen);
#endif

  if (colormap)
    {
      gtk_widget_set_default_colormap (colormap);
    }
  else
    {
      g_warning ("The screen does not support the alpha channel\n");
      visual = gdk_screen_get_rgba_visual (screen);

      if (visual == NULL)
	{
	  g_warning ("The screen does not support the rgba visual!\n");
#ifndef _WIN32
	  exit (EXIT_FAILURE);
#endif
	}

    }

}


/* Parse the command line in the standard getopt way. */
static CommandLine *
parse_options (gint argc,
	       char *argv[])
{
  CommandLine *commandline = g_malloc ((gsize) sizeof (CommandLine));

  commandline->position = EAST;
  commandline->debug = FALSE;
  commandline->iwb_filename = NULL;
  commandline->decorated=FALSE;

  /* Getopt_long stores the option index here. */
  while (1)
    {
      gint c;
      static struct option long_options[] =
	{
	  /* These options set a flag. */
	  {"help", no_argument,       0, 'h'},
          {"decorated", no_argument,  0, 'd'},
	  {"verbose", no_argument,    0, 'V'},
          {"version", no_argument,    0, 'v'},
	  /* 
           * These options don't set a flag.
	   * We distinguish them by their indices.
           */
	  {"gravity", required_argument, 0, 'g'},
	  {0, 0, 0, 0}
	};

      gint option_index = 0;
      c = getopt_long (argc,
		       argv,
		       "hdvVg:",
		       long_options,
		       &option_index);

      /* Detect the end of the options. */
      if (c == -1)
	break;

      switch (c)
	{
	case 'h':
	  print_help ();
	  break;
        case 'v':
	  print_version ();
	  break;
        case 'd':
	  commandline->decorated=TRUE;
	  break;
	case 'V':
	  commandline->debug=TRUE;
	  break;
	case 'g':
	  if (g_strcmp0 (optarg, "east") == 0)
	    {
	      commandline->position = EAST;
	    }
	  else if (g_strcmp0 (optarg, "west") == 0)
	    {
	      commandline->position = WEST;
	    }
          else if (g_strcmp0 (optarg, "north") == 0)
	    {
	      commandline->position = NORTH;
	    }
          else if (g_strcmp0 (optarg, "south") == 0)
	    {
	      commandline->position = SOUTH;
	    }
	  else
	    {
	      print_help ();
	    }
	  break;
	default:
	  print_help ();
	  break;
	} 
    }

  if (optind<argc)
    {
      commandline->iwb_filename = argv[optind];
    }

  return commandline;
}


/* Enable the localization support with gettext. */
static void
enable_localization_support ()
{
#ifdef ENABLE_NLS
  setlocale (LC_ALL, "");
  bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
  textdomain (GETTEXT_PACKAGE);
#endif

  gtk_set_locale ();
}


/* Create a shorcut to the workspace on the desktop. */
static void
create_workspace_shortcut (gchar *workspace_dir)
{
  gchar *desktop_entry_filename = g_strdup_printf ("%s%s%s_workspace",
						   get_desktop_dir (),
						   G_DIR_SEPARATOR_S,
						   PACKAGE_NAME);

#ifdef _WIN32
  windows_create_link (workspace_dir,
		       desktop_entry_filename,
		       "%SystemRoot%\\system32\\imageres.dll",
		       123);

#else
  xdg_create_link (workspace_dir , desktop_entry_filename, "folder-documents");
#endif
  g_free (desktop_entry_filename);
}


/* Configure the workspace. */
static gchar *
configure_workspace (gchar *project_name)
{
  /* The workspace directory is set in the documents ardesia folder. */
  gchar *workspace_dir = g_build_filename (get_documents_dir (),
					   PACKAGE_NAME,
					   (gchar *) 0);

  create_workspace_shortcut (workspace_dir);

  return workspace_dir;
}


/* Create the default project dir under the workspace_dir. */
static gchar *
create_default_project_dir (gchar *workspace_dir,
			    gchar *project_name)
{
  gchar *project_dir = g_build_filename (workspace_dir, project_name, (gchar *) 0);

  if (!g_file_test (project_dir, G_FILE_TEST_EXISTS)) 
    {

      if (g_mkdir_with_parents (project_dir, 0700)==-1)
        {
	  g_warning ("Unable to create folder %s\n", project_dir);
        }

    }

  return project_dir;
}


/* This is the starting point of the program. */
int
main (int argc,
      char *argv[])
{
  CommandLine *commandline = (CommandLine *) NULL;
  gchar       *project_name = (gchar *) NULL;
  gchar       *project_dir = (gchar *) NULL;
  gchar       *iwb_filename = (gchar *) NULL;
  GtkWidget   *background_window = (GtkWidget *) NULL;
  GtkWidget   *annotation_window = (GtkWidget *) NULL;
  GtkWidget   *ardesia_bar_window = (GtkWidget *) NULL;
  GSList      *artifact_list = (GSList *) NULL;

  /* Enable the localization support with gettext. */
  enable_localization_support ();

#ifdef HAVE_BACKTRACE
#ifdef HAVE_LIBSIGSEGV
  glibc_backtrace_register ();
#endif
#endif
#ifdef _WIN32
  windows_backtrace_register ();
#endif


  gtk_init (&argc, &argv);
	
  set_the_best_colormap ();

  /*
   * Uncomment this and the create_segmentation_fault function
   * to create a segmentation fault
   */
  //create_segmentation_fault ();

  commandline = parse_options (argc, argv);
	
  if (commandline->iwb_filename)
    {
      gint init_pos = -1;
      gint end_pos = -1;


      if (g_path_is_absolute (commandline->iwb_filename))
	{
	  iwb_filename = g_strdup (commandline->iwb_filename);
	}
      else
	{
	  gchar *dir = g_get_current_dir ();
	  iwb_filename = g_build_filename (dir, commandline->iwb_filename, (gchar *) 0);
	  g_free (dir);
	}

      if (!file_exists(iwb_filename))
        {
          printf("No such file %s\n", iwb_filename);          
          exit (EXIT_FAILURE);          
        }

      init_pos = g_substrlastpos (iwb_filename, G_DIR_SEPARATOR_S);
      end_pos  = g_substrlastpos (iwb_filename, ".");
      project_name = g_substr (iwb_filename, init_pos+1, end_pos-1);
      project_dir = g_substr (iwb_filename, 0, init_pos-1);
    }
  else
    {
      gchar *workspace_dir = (gchar *) NULL;

      /* Show the project name wizard. */
      project_name = start_project_dialog ((GtkWindow *) NULL);
      workspace_dir = configure_workspace (project_name);
      project_dir = create_default_project_dir (workspace_dir, project_name);
      g_free (workspace_dir);
    }

  set_project_name (project_name);
  set_project_dir (project_dir);
  set_iwb_filename (iwb_filename);


  background_window = create_background_window ();

  if (background_window == NULL)
    {
      destroy_background_window ();
      g_free (commandline);
      exit (EXIT_FAILURE);
    }

  gtk_window_set_keep_above (GTK_WINDOW (background_window), TRUE);
  gtk_widget_show (background_window);
  
  set_background_window (background_window);
  
  /* Initialize the annotation window. */
  annotate_init (background_window, iwb_filename, commandline->debug);

  annotation_window = get_annotation_window ();

  if (annotation_window == NULL)
    {
      annotate_quit ();
      destroy_background_window ();
      g_free (commandline);
      exit (EXIT_FAILURE);
    }

  /* Postcondition: the annotation window is valid. */
  gtk_window_set_keep_above (GTK_WINDOW (annotation_window), TRUE);

  gtk_widget_show (annotation_window);

  ardesia_bar_window = create_bar_window (commandline, annotation_window);

  if (ardesia_bar_window == NULL)
    {
      annotate_quit ();
      destroy_background_window ();
      g_free (commandline);
      exit (EXIT_FAILURE);
    }

  gtk_window_set_keep_above (GTK_WINDOW (ardesia_bar_window), TRUE);
  gtk_widget_show (ardesia_bar_window);
  gtk_main ();

  artifact_list = get_artifacts ();

  if (artifact_list)
    {
      start_share_dialog ((GtkWindow *) NULL);
      free_artifacts ();
    }

  g_free (project_name);

  remove_dir_if_empty(project_dir);

  g_free (project_dir);
  
  g_free (iwb_filename);

  g_free (commandline);

  return 0;
}


